File: problem1.go
// Time Complexity : O(n!)
// Space Complexity : O(nxn)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
var result [][] string
var grid [][]bool
var m int

func solveNQueens(n int) [][]string {
    if n == 0{
        return result
    }
    
    result = [][]string{}
    grid = make([][]bool,n)
    for i:=0;i<n;i++{
        grid[i] = make([]bool,n)
    }
    
    m = n
    
    backtrack(0)
    
    return result
}


func backtrack(row int){
    
    // base
    if row == m{
        newlist := []string{}
        for i:=0;i<m;i++{
            word := ""
            for j:=0;j<m;j++{
                if grid[i][j] == true{
                    word += "Q"
                }else{
                    word+="."
                }
                
            }
            newlist = append(newlist,word)
            
        }
        
        result = append(result,newlist)
        
        return
    }
    
    
    //logic 
    for i:=0;i<m;i++{
        if isSafe(row,i){
            grid[row][i] = true
            
            backtrack(row+1)
            
            grid[row][i] = false
        }
    }
}

func isSafe(r,c int) bool{
    
    for i:=r-1;i>=0;i--{
        if grid[i][c]{
            return false
        }      
    }
    
    i := r
    j := c
    
    for i >=0 && j >=0{
        if grid[i][j]{
            return false
        } 
        
        i--
        j--
    }
    
    i = r
    j = c
    
    for i >=0 && j <m {
        if grid[i][j]{
            return false
        } 
        
        i--
        j++
    }
    
    return true
}

File: Problem1.cs
public class NQueens
    {
        // Time Complexity : O(n!)- where n is the number of queens
        // Space Complexity : O(n^2) -  for the grid used in computation n * n.
        // Did this code successfully run on Leetcode : Yes
        // Any problem you faced while coding this : No

        IList<IList<string>> result;
        bool[,] grid;

        public IList<IList<string>> SolveNQueens(int n)
        {
            result = new List<IList<string>>();
            grid = new bool[n,n];
            backtrack(0, n);
            return result;
        }

        private void backtrack(int r, int  n)
        {
            //base case
            if(r == n)
            {
                List<string> list = new List<string>();
                for(int i = 0; i < n; i++)
                {
                    StringBuilder strRow = new StringBuilder();
                    for(int j = 0; j < n; j++)
                    {
                        if (grid[i,j])
                        {
                            strRow.Append("Q");
                        }
                        else
                        {
                            strRow.Append(".");
                        }
                    }
                    list.Add(strRow.ToString());
                }

                result.Add(list);
            }

            //logic
            for(int c = 0; c < n; c++)
            {
                if(IsSafe(r, c, n))
                {
                    //action
                    grid[r, c] = true;

                    //recurse
                    backtrack(r + 1, n);

                    //backtrack
                    grid[r, c] = false;
                }
            }
        }

        private bool IsSafe(int r, int c, int n)
        {
            int i = r;
            int j = c;
            //column up
            for (i = 0; i < r; i++)
            {
                if (grid[i, c])
                    return false;
            }
            //diagonal up left
            
            while(i >= 0 && j >=0)
            {
                if(grid[i, j])
                    return false;
                i--;
                j--;
            }
            //diagonal up right
            i = r;
            j = c;
            while(i >= 0 && j < n)
            {
                if (grid[i, j])
                    return false;
                i--;
                j++;
            }
            return true;
        }
    }
